File: C++/sort/quick_sort.cpp
// Copyright 2019 Aman Mehara
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "merge_sort.h"

#include <algorithm>
#include <vector>

namespace mehara::sort {

int partition(std::vector<double>& array, int begin, int end) {
    auto pivot = array[end];
    auto i = begin;
    for (auto j = i; j < end; j++) {
        if (array[j] < pivot) {
            std::swap(array[i], array[j]);
            i += 1;
        }
    }
    std::swap(array[i], array[end]);
    return i;
}

void quick_sort(std::vector<double>& array, int begin, int end) {
    if (begin <= end) {
        auto _partition = partition(array, begin, end);
        quick_sort(array, begin, _partition - 1);
        quick_sort(array, _partition + 1, end);
    }
}

void quick_sort(std::vector<double>& array) {
    auto begin = 0;
    auto end = array.size() - 1;

    quick_sort(array, begin, end);
}

} // namespace mehara::sort

File: Go/QuickSort.go
/*Copyright 2020 Anil Kumar Teegala
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
    
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Code for Quick Sort in Go

package main
 
import (
    "fmt"
    "math/rand"
    "time"
)
 
func main() {
 
    slice := generateSlice(20)
    fmt.Println("\n--- Unsorted --- \n\n", slice)
    quicksort(slice)
    fmt.Println("\n--- Sorted ---\n\n", slice, "\n")
}
 
// Generates a slice of size, size filled with random numbers
func generateSlice(size int) []int {
 
    slice := make([]int, size, size)
    rand.Seed(time.Now().UnixNano())
    for i := 0; i < size; i++ {
        slice[i] = rand.Intn(999) - rand.Intn(999)
    }
    return slice
}
  
func quicksort(a []int) []int {
    if len(a) < 2 {
        return a
    }
      
    left, right := 0, len(a)-1
      
    pivot := rand.Int() % len(a)
      
    a[pivot], a[right] = a[right], a[pivot]
      
    for i, _ := range a {
        if a[i] < a[right] {
            a[left], a[i] = a[i], a[left]
            left++
        }
    }
      
    a[left], a[right] = a[right], a[left]
      
    quicksort(a[:left])
    quicksort(a[left+1:])
      
    return a
}
